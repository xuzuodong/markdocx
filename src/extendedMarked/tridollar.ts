import { marked } from 'marked'

/**
 * custom syntax
 * i.e. $$$ Some text here ... $$$
 */
export const tridollar: marked.TokenizerExtension = {
    name: 'tridollar',
    level: 'block',
    tokenizer(src: string) {
        const rule = /^\${3}(.+)\${3}(:?\n|$)/
        const match = rule.exec(src)

        if (match) {
            const token = {
                type: 'tridollar',
                raw: match[0],
                text: match[1].trim(),
                tokens: [],
            }
            return token
        }
    },
}
