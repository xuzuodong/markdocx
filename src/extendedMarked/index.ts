import { marked } from 'marked'
import { tridollar } from './tridollar'

marked.use({ extensions: [tridollar] })

export * from 'marked'
