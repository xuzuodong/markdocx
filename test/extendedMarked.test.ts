import fs from 'fs'
import { marked } from '../src/extendedMarked/index'
import { test, expect } from 'vitest'

test('extended marked', () => {
    const src = fs.readFileSync('./examples/test.md').toString()
    expect(marked.Lexer.lex(src)).toMatchSnapshot()
})
