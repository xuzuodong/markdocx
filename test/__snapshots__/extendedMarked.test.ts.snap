// Vitest Snapshot v1

exports[`extended marked 1`] = `
[
  {
    "depth": 1,
    "raw": "# 科学和人文谁更有意义

",
    "text": "科学和人文谁更有意义",
    "tokens": [
      {
        "raw": "科学和人文谁更有意义",
        "text": "科学和人文谁更有意义",
        "type": "text",
      },
    ],
    "type": "heading",
  },
  {
    "raw": "本人也是经过了深思熟虑，**在每个日日夜夜**思考这个问题。问题的关键究竟为何？科学和人文谁更有意义，发生了会如何，不发生又会如何。 科学和人文谁更有意义，到底应该如何实现。*就我个人来说，**科学和人文**谁更有意义对我的意义，不能不说非常重大*。 而这些并不是完全重要，更加重要的问题是， 我们都知道，只要有意义，那么就必须慎重考虑。 科学和人文谁更有意义，发生了会如何，不发生又会如何。吉格·金克拉说过一句富有哲理的话，如果你能做梦，你就能实现它。带着这句话，我们还要更加慎重的审视这个问题： 吕凯特说过一句富有哲理的话，生命不可能有两次，但许多人连一次也不善于度过。这似乎解答了我的疑惑。 所谓科学和人文谁更有意义，关键是科学和人文谁更有意义需要如何写。 从这个角度来看， 我们一般认为，抓住了问题的关键，其他一切则会迎刃而解。",
    "text": "本人也是经过了深思熟虑，**在每个日日夜夜**思考这个问题。问题的关键究竟为何？科学和人文谁更有意义，发生了会如何，不发生又会如何。 科学和人文谁更有意义，到底应该如何实现。*就我个人来说，**科学和人文**谁更有意义对我的意义，不能不说非常重大*。 而这些并不是完全重要，更加重要的问题是， 我们都知道，只要有意义，那么就必须慎重考虑。 科学和人文谁更有意义，发生了会如何，不发生又会如何。吉格·金克拉说过一句富有哲理的话，如果你能做梦，你就能实现它。带着这句话，我们还要更加慎重的审视这个问题： 吕凯特说过一句富有哲理的话，生命不可能有两次，但许多人连一次也不善于度过。这似乎解答了我的疑惑。 所谓科学和人文谁更有意义，关键是科学和人文谁更有意义需要如何写。 从这个角度来看， 我们一般认为，抓住了问题的关键，其他一切则会迎刃而解。",
    "tokens": [
      {
        "raw": "本人也是经过了深思熟虑，",
        "text": "本人也是经过了深思熟虑，",
        "type": "text",
      },
      {
        "raw": "**在每个日日夜夜**",
        "text": "在每个日日夜夜",
        "tokens": [
          {
            "raw": "在每个日日夜夜",
            "text": "在每个日日夜夜",
            "type": "text",
          },
        ],
        "type": "strong",
      },
      {
        "raw": "思考这个问题。问题的关键究竟为何？科学和人文谁更有意义，发生了会如何，不发生又会如何。 科学和人文谁更有意义，到底应该如何实现。",
        "text": "思考这个问题。问题的关键究竟为何？科学和人文谁更有意义，发生了会如何，不发生又会如何。 科学和人文谁更有意义，到底应该如何实现。",
        "type": "text",
      },
      {
        "raw": "*就我个人来说，**科学和人文**谁更有意义对我的意义，不能不说非常重大*",
        "text": "就我个人来说，**科学和人文**谁更有意义对我的意义，不能不说非常重大",
        "tokens": [
          {
            "raw": "就我个人来说，",
            "text": "就我个人来说，",
            "type": "text",
          },
          {
            "raw": "**科学和人文**",
            "text": "科学和人文",
            "tokens": [
              {
                "raw": "科学和人文",
                "text": "科学和人文",
                "type": "text",
              },
            ],
            "type": "strong",
          },
          {
            "raw": "谁更有意义对我的意义，不能不说非常重大",
            "text": "谁更有意义对我的意义，不能不说非常重大",
            "type": "text",
          },
        ],
        "type": "em",
      },
      {
        "raw": "。 而这些并不是完全重要，更加重要的问题是， 我们都知道，只要有意义，那么就必须慎重考虑。 科学和人文谁更有意义，发生了会如何，不发生又会如何。吉格·金克拉说过一句富有哲理的话，如果你能做梦，你就能实现它。带着这句话，我们还要更加慎重的审视这个问题： 吕凯特说过一句富有哲理的话，生命不可能有两次，但许多人连一次也不善于度过。这似乎解答了我的疑惑。 所谓科学和人文谁更有意义，关键是科学和人文谁更有意义需要如何写。 从这个角度来看， 我们一般认为，抓住了问题的关键，其他一切则会迎刃而解。",
        "text": "。 而这些并不是完全重要，更加重要的问题是， 我们都知道，只要有意义，那么就必须慎重考虑。 科学和人文谁更有意义，发生了会如何，不发生又会如何。吉格·金克拉说过一句富有哲理的话，如果你能做梦，你就能实现它。带着这句话，我们还要更加慎重的审视这个问题： 吕凯特说过一句富有哲理的话，生命不可能有两次，但许多人连一次也不善于度过。这似乎解答了我的疑惑。 所谓科学和人文谁更有意义，关键是科学和人文谁更有意义需要如何写。 从这个角度来看， 我们一般认为，抓住了问题的关键，其他一切则会迎刃而解。",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  {
    "raw": "

",
    "type": "space",
  },
  {
    "depth": 2,
    "raw": "## 要想清楚，科学和人文谁更有意义，到底是一种怎么样的存在

",
    "text": "要想清楚，科学和人文谁更有意义，到底是一种怎么样的存在",
    "tokens": [
      {
        "raw": "要想清楚，科学和人文谁更有意义，到底是一种怎么样的存在",
        "text": "要想清楚，科学和人文谁更有意义，到底是一种怎么样的存在",
        "type": "text",
      },
    ],
    "type": "heading",
  },
  {
    "raw": "那么， 每个人都不得不面对这些问题。 在面对这种问题时， 对我个人而言，科学和人文谁更有意义不仅仅是一个重大的事件，**还可能会改变我的人生**。 对我个人而言，科学和人文谁更有意义不仅仅是一个重大的事件，还可能会改变我的人生。 所谓科学和人文谁更有意义，关键是科学和人文谁更有意义需要如何写。 这种事实对本人来说意义重大，相信对这个世界也是有一定意义的。 那么， 而这些并不是完全重要，更加重要的问题是， 每个人都不得不面对这些问题。",
    "text": "那么， 每个人都不得不面对这些问题。 在面对这种问题时， 对我个人而言，科学和人文谁更有意义不仅仅是一个重大的事件，**还可能会改变我的人生**。 对我个人而言，科学和人文谁更有意义不仅仅是一个重大的事件，还可能会改变我的人生。 所谓科学和人文谁更有意义，关键是科学和人文谁更有意义需要如何写。 这种事实对本人来说意义重大，相信对这个世界也是有一定意义的。 那么， 而这些并不是完全重要，更加重要的问题是， 每个人都不得不面对这些问题。",
    "tokens": [
      {
        "raw": "那么， 每个人都不得不面对这些问题。 在面对这种问题时， 对我个人而言，科学和人文谁更有意义不仅仅是一个重大的事件，",
        "text": "那么， 每个人都不得不面对这些问题。 在面对这种问题时， 对我个人而言，科学和人文谁更有意义不仅仅是一个重大的事件，",
        "type": "text",
      },
      {
        "raw": "**还可能会改变我的人生**",
        "text": "还可能会改变我的人生",
        "tokens": [
          {
            "raw": "还可能会改变我的人生",
            "text": "还可能会改变我的人生",
            "type": "text",
          },
        ],
        "type": "strong",
      },
      {
        "raw": "。 对我个人而言，科学和人文谁更有意义不仅仅是一个重大的事件，还可能会改变我的人生。 所谓科学和人文谁更有意义，关键是科学和人文谁更有意义需要如何写。 这种事实对本人来说意义重大，相信对这个世界也是有一定意义的。 那么， 而这些并不是完全重要，更加重要的问题是， 每个人都不得不面对这些问题。",
        "text": "。 对我个人而言，科学和人文谁更有意义不仅仅是一个重大的事件，还可能会改变我的人生。 所谓科学和人文谁更有意义，关键是科学和人文谁更有意义需要如何写。 这种事实对本人来说意义重大，相信对这个世界也是有一定意义的。 那么， 而这些并不是完全重要，更加重要的问题是， 每个人都不得不面对这些问题。",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  {
    "raw": "

",
    "type": "space",
  },
  {
    "depth": 3,
    "raw": "### 问题的关键究竟为何？

",
    "text": "问题的关键究竟为何？",
    "tokens": [
      {
        "raw": "问题的关键究竟为何？",
        "text": "问题的关键究竟为何？",
        "type": "text",
      },
    ],
    "type": "heading",
  },
  {
    "items": [
      {
        "checked": undefined,
        "loose": false,
        "raw": "- 卡耐基说过一句富有哲理的话，我们若已接受最坏的，就再没有什么损失。
",
        "task": false,
        "text": "卡耐基说过一句富有哲理的话，我们若已接受最坏的，就再没有什么损失。",
        "tokens": [
          {
            "raw": "卡耐基说过一句富有哲理的话，我们若已接受最坏的，就再没有什么损失。",
            "text": "卡耐基说过一句富有哲理的话，我们若已接受最坏的，就再没有什么损失。",
            "tokens": [
              {
                "raw": "卡耐基说过一句富有哲理的话，我们若已接受最坏的，就再没有什么损失。",
                "text": "卡耐基说过一句富有哲理的话，我们若已接受最坏的，就再没有什么损失。",
                "type": "text",
              },
            ],
            "type": "text",
          },
        ],
        "type": "list_item",
      },
      {
        "checked": undefined,
        "loose": false,
        "raw": "- 这启发了我， 科学和人文谁更有意义因何而发生？
",
        "task": false,
        "text": "这启发了我， 科学和人文谁更有意义因何而发生？",
        "tokens": [
          {
            "raw": "这启发了我， 科学和人文谁更有意义因何而发生？",
            "text": "这启发了我， 科学和人文谁更有意义因何而发生？",
            "tokens": [
              {
                "raw": "这启发了我， 科学和人文谁更有意义因何而发生？",
                "text": "这启发了我， 科学和人文谁更有意义因何而发生？",
                "type": "text",
              },
            ],
            "type": "text",
          },
        ],
        "type": "list_item",
      },
      {
        "checked": undefined,
        "loose": false,
        "raw": "- 了解清楚科学和人文谁更有意义到底是一种怎么样的存在，是解决一切问题的关键。 
",
        "task": false,
        "text": "了解清楚科学和人文谁更有意义到底是一种怎么样的存在，是解决一切问题的关键。 ",
        "tokens": [
          {
            "raw": "了解清楚科学和人文谁更有意义到底是一种怎么样的存在，是解决一切问题的关键。 ",
            "text": "了解清楚科学和人文谁更有意义到底是一种怎么样的存在，是解决一切问题的关键。 ",
            "tokens": [
              {
                "raw": "了解清楚科学和人文谁更有意义到底是一种怎么样的存在，是解决一切问题的关键。 ",
                "text": "了解清楚科学和人文谁更有意义到底是一种怎么样的存在，是解决一切问题的关键。 ",
                "type": "text",
              },
            ],
            "type": "text",
          },
        ],
        "type": "list_item",
      },
      {
        "checked": undefined,
        "loose": false,
        "raw": "- 叔本华曾经提到过，普通人只想到如何度过时间，有才能的人设法利用时间。我希望诸位也能好好地体会这句话。",
        "task": false,
        "text": "叔本华曾经提到过，普通人只想到如何度过时间，有才能的人设法利用时间。我希望诸位也能好好地体会这句话。",
        "tokens": [
          {
            "raw": "叔本华曾经提到过，普通人只想到如何度过时间，有才能的人设法利用时间。我希望诸位也能好好地体会这句话。",
            "text": "叔本华曾经提到过，普通人只想到如何度过时间，有才能的人设法利用时间。我希望诸位也能好好地体会这句话。",
            "tokens": [
              {
                "raw": "叔本华曾经提到过，普通人只想到如何度过时间，有才能的人设法利用时间。我希望诸位也能好好地体会这句话。",
                "text": "叔本华曾经提到过，普通人只想到如何度过时间，有才能的人设法利用时间。我希望诸位也能好好地体会这句话。",
                "type": "text",
              },
            ],
            "type": "text",
          },
        ],
        "type": "list_item",
      },
    ],
    "loose": false,
    "ordered": false,
    "raw": "- 卡耐基说过一句富有哲理的话，我们若已接受最坏的，就再没有什么损失。
- 这启发了我， 科学和人文谁更有意义因何而发生？
- 了解清楚科学和人文谁更有意义到底是一种怎么样的存在，是解决一切问题的关键。 
- 叔本华曾经提到过，普通人只想到如何度过时间，有才能的人设法利用时间。我希望诸位也能好好地体会这句话。",
    "start": "",
    "type": "list",
  },
  {
    "raw": "

",
    "type": "space",
  },
  {
    "raw": "Have some English words here, English words can have different font other than Chinese words.",
    "text": "Have some English words here, English words can have different font other than Chinese words.",
    "tokens": [
      {
        "raw": "Have some English words here, English words can have different font other than Chinese words.",
        "text": "Have some English words here, English words can have different font other than Chinese words.",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  {
    "raw": "

",
    "type": "space",
  },
  {
    "align": [
      "left",
      "center",
      "right",
    ],
    "header": [
      {
        "text": "Syntax",
        "tokens": [
          {
            "raw": "Syntax",
            "text": "Syntax",
            "type": "text",
          },
        ],
      },
      {
        "text": "Description",
        "tokens": [
          {
            "raw": "Description",
            "text": "Description",
            "type": "text",
          },
        ],
      },
      {
        "text": "Test Text",
        "tokens": [
          {
            "raw": "Test Text",
            "text": "Test Text",
            "type": "text",
          },
        ],
      },
    ],
    "raw": "| Syntax    | Description |   Test Text |
| :-------- | :---------: | ----------: |
| Header    |    Title    | Here's this |
| Paragraph |    Text     |    And more |

",
    "rows": [
      [
        {
          "text": "Header",
          "tokens": [
            {
              "raw": "Header",
              "text": "Header",
              "type": "text",
            },
          ],
        },
        {
          "text": "Title",
          "tokens": [
            {
              "raw": "Title",
              "text": "Title",
              "type": "text",
            },
          ],
        },
        {
          "text": "Here's this",
          "tokens": [
            {
              "raw": "Here's this",
              "text": "Here&#39;s this",
              "type": "text",
            },
          ],
        },
      ],
      [
        {
          "text": "Paragraph",
          "tokens": [
            {
              "raw": "Paragraph",
              "text": "Paragraph",
              "type": "text",
            },
          ],
        },
        {
          "text": "Text",
          "tokens": [
            {
              "raw": "Text",
              "text": "Text",
              "type": "text",
            },
          ],
        },
        {
          "text": "And more",
          "tokens": [
            {
              "raw": "And more",
              "text": "And more",
              "type": "text",
            },
          ],
        },
      ],
    ],
    "type": "table",
  },
  {
    "raw": "**我是整段加粗文字！**",
    "text": "**我是整段加粗文字！**",
    "tokens": [
      {
        "raw": "**我是整段加粗文字！**",
        "text": "我是整段加粗文字！",
        "tokens": [
          {
            "raw": "我是整段加粗文字！",
            "text": "我是整段加粗文字！",
            "type": "text",
          },
        ],
        "type": "strong",
      },
    ],
    "type": "paragraph",
  },
  {
    "raw": "

",
    "type": "space",
  },
  {
    "raw": "![img caption here](./testimg.jpg)",
    "text": "![img caption here](./testimg.jpg)",
    "tokens": [
      {
        "href": "./testimg.jpg",
        "raw": "![img caption here](./testimg.jpg)",
        "text": "img caption here",
        "title": null,
        "type": "image",
      },
    ],
    "type": "paragraph",
  },
]
`;
